vowpalwabbit/learner.h

```diff
@@ -678,5 +690,212 @@ void multiline_learn_or_predict(multi_learner& base, multi_ex& examples, const u
+  }
+
+  FluentBuilderT& set_save_load(void (*sl)(DataT&, io_buf&, bool, bool))
+  {
+    _learner->save_load_fd.save_load_f = (save_load_data::fn)sl;
+    _learner->save_load_fd.data = _learner->learn_fd.data;
+    _learner->save_load_fd.base = _learner->learn_fd.base;
+    return *static_cast<FluentBuilderT*>(this);
+  }
+
+  FluentBuilderT& set_finish(void (*f)(DataT&))
+  {
+    _learner->finisher_fd = func_data(_learner->learn_fd.data, _learner->learn_fd.base, (finish_fptr_type)(f));
+    return *static_cast<FluentBuilderT*>(this);
+  }
+
+  FluentBuilderT& set_end_pass(void (*f)(DataT&))
+  {
+    _learner->end_pass_fd = func_data(_learner->learn_fd.data, _learner->learn_fd.base, (func_data::fn)f);
+    return *static_cast<FluentBuilderT*>(this);
+  }
+
+  FluentBuilderT& set_end_examples(void (*f)(DataT&))
+  {
+    _learner->end_examples_fd = func_data(_learner->learn_fd.data, _learner->learn_fd.base, (func_data::fn)f);
+    return *static_cast<FluentBuilderT*>(this);
+  }
+
+  FluentBuilderT& set_init_driver(void (*f)(DataT&))
+  {
+    _learner->init_fd = func_data(_learner->learn_fd.data, _learner->learn_fd.base, (func_data::fn)f);
+    return *static_cast<FluentBuilderT*>(this);
+  }
+
+  FluentBuilderT& set_finish_example(void (*f)(vw& all, DataT&, ExampleT&))
+  {
+    _learner->finish_example_fd.data = _learner->learn_fd.data;
+    _learner->finish_example_fd.finish_example_f = (end_fptr_type)(f);
+    return *static_cast<FluentBuilderT*>(this);
+  }
+};
+
+template <class DataT, class ExampleT, class BaseLearnerT>
+struct reduction_learner_builder
+    : public common_learner_builder<reduction_learner_builder<DataT, ExampleT, BaseLearnerT>, DataT, ExampleT,
+          BaseLearnerT>
+{
+  reduction_learner_builder(std::unique_ptr<DataT>&& data, BaseLearnerT* base, const std::string& name)
+      // NOTE: This is a copy of the base! The purpose is to copy all of the
+      // function data objects so that if this reduction does not define a function such as
+      // save_load then calling save_load on this object will essentially result in forwarding the
+      // call the next reduction that actually implements it.
+      : common_learner_builder<reduction_learner_builder<DataT, ExampleT, BaseLearnerT>, DataT, ExampleT, BaseLearnerT>(
+            new learner<DataT, ExampleT>(*reinterpret_cast<learner<DataT, ExampleT>*>(base)), std::move(data), name)
+  {
+    this->_learner->learn_fd.base = make_base(*base);
+    this->_learner->learn_fd.data = this->_learner->learner_data.get();
+    this->_learner->sensitivity_fd.sensitivity_f = (sensitivity_data::fn)recur_sensitivity;
+    this->_learner->finisher_fd.data = this->_learner->learner_data.get();
+    this->_learner->finisher_fd.base = make_base(*base);
+    this->_learner->finisher_fd.func = (func_data::fn)noop;
+
+    set_params_per_weight(1);
+
+    this->_learner->pred_type = base->pred_type;
+    // TODO add label type as something learner knows about itself.
```

Comment:
is this bit of code commented out on purpose?

Reply:
Yes, learner does not yet support having the label supplied